### PR TITLE
Added caching and retry policy

### PR DESCRIPTION
**Pull request by PeterDowdy**
Hi, I put some retry and caching semantics on your web requests using Polly. The gist of it is that all web requests now retry thrice and they will also cache their responses. The cache is based on what the uri is, since all requests are GETs, and the retention period is variable. Caching uses the local temp directory, with a file scoped to only this user/process, so it persists between invocations but should not be accessible by other processes. Caching only occurs if `--use-cache` is on. Retry always happens.

image

### CODE REVIEW

**Comment by PeterDowdy** [comment]
@rgwood any feedback on this PR?

**Comment by rgwood** [comment]
Hey, thanks that’s awesome! For some reason I didn’t get a notification for your original PR, only your comment just now.

I’ll take a closer look tomorrow but from the description that sounds great.

**Comment by rgwood** [comment]
The cache is super handy, I wasn't aware that Polly had that kind of functionality.

Favour to ask, can you manually confirm that the new functionality works with [trimming](https://docs.microsoft.com/en-us/dotnet/core/deploying/trimming/trimming-options)? There are publish profiles with trimming enabled in the RezoningScraper project, they work with the VS publish feature or the dotnet CLI [like this](https://github.com/rgwood/RezoningScraper/blob/master/publish-all.sh#L3).

File: RezoningScraper/CacheManager.cs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+using static Spectre.Console.AnsiConsole;
+
+namespace RezoningScraper
+{
+    public class CacheManager<TResult> : IAsyncCacheProvider<TResult>
+    {
+        record CachedRecord<T>(DateTime Ttl, T Record);
+        public async Task PutAsync(string key, TResult value, Ttl ttl, CancellationToken ct, bool continueOnCapturedContext)
+        {
+            using IsolatedStorageFile isoStore = IsolatedStorageFile.GetStore(IsolatedStorageScope.User | IsolatedStorageScope.Domain | IsolatedStorageScope.Assembly, null, null);
```
Comment (rgwood) [inline RezoningScraper/CacheManager.cs:15]: This is interesting, I haven't used [isolated storage](https://docs.microsoft.com/en-us/dotnet/standard/io/isolated-storage) before. This bit from the docs gives me pause:

> This is a legacy component primarily designed for now-deprecated Code Access Security scenarios.

The docs are also unclear on how/whether it works on non-Windows platforms.

I'm OK using it if it does the trick, but I'd like to spend a little bit of time understanding it better first. An alternative might be to dump it in the database.

File: RezoningScraper/API.cs [hunk 2]
```diff
@@ -12,10 +13,14 @@ public static class API
...
 	/// <returns>An async enumerable of projects (because the API is paginated)</returns>
-	public static async IAsyncEnumerable<Project> GetAllProjects(string jwt)
+	public static async IAsyncEnumerable<Project> GetAllProjects(string jwt, bool useCache = false)
 	{
+
+			IAsyncPolicy<Projects> cachePolicy = useCache
+				? Policy.CacheAsync<Projects>(new CacheManager<Projects>(), TimeSpan.FromHours(1))
+				: Policy.NoOpAsync<Projects>();
+			var client = new HttpClient();
 		string startUrl = $"https://shapeyourcity.ca/api/v2/projects?per_page={ResultsPerPage}";
```
Comment (rgwood) [inline RezoningScraper/API.cs:23]: Mind fixing up the formatting here? New code is indented more than the existing code

File: RezoningScraper/CacheManager.cs (hunk 1 above)
Comment (PeterDowdy) [reply RezoningScraper/CacheManager.cs:15]: Isolated storage is not an API with a future, so if you are comfortable putting the cache in SQLite, I'll move it there.

**Comment by PeterDowdy** [comment]
Yep, I'll verify the Release build and post evidence once complete.

File: RezoningScraper/CacheManager.cs (hunk 1 above)
Comment (rgwood) [reply RezoningScraper/CacheManager.cs:15]: Sounds good.

I don't have a strong opinion on what the schema should look like; could be in its own table, or we could create a shared cache table for tokens and API responses.

File: RezoningScraper/API.cs (hunk 2 above)
Comment (PeterDowdy) [reply RezoningScraper/API.cs:23]: Yep, I've been writing a lot of python professionally and have been spoiled by black auto-formatting everything on save.

**Comment by PeterDowdy** [comment]
image

Here's the application running with a trimmed executable.

File: RezoningScraper/CacheManager.cs (hunk 1 above)
Comment (PeterDowdy) [reply RezoningScraper/CacheManager.cs]: The cache table is a generic key-value cache with customizable expiry so you could move the token in there if you wanted.

File: RezoningScraper/CacheManager.cs [hunk 3]
```diff
@@ -0,0 +1,76 @@
+﻿using Polly.Caching;
+using System.IO.IsolatedStorage;
```
Comment (rgwood) [inline RezoningScraper/CacheManager.cs:2]: This using should be unnecessary now

File: RezoningScraper/TokenHelper.cs [hunk 4]
```diff
@@ -20,23 +21,29 @@ internal static async Task<Token> GetTokenFromDbOrWebsite(SqliteConnection db)
         else
         {
             // TODO: add retries, this page seems unreliable
```
Comment (rgwood) [inline RezoningScraper/TokenHelper.cs:23]: This comment's no longer valid (thanks)!

**Comment by rgwood** [comment]
Left some minor comments. If you want to bump the project version I can merge this and release it later today. Thanks!

**Comment by PeterDowdy** [comment]
Done

**Comment by rgwood** [comment]
LGTM. Thanks again!

**Pull request merged**